### DISCUSSION TITLE
sqlalchemy.types.Text and portability

### DISCUSSION THREAD

**Discussion by mbeijen** [sqlalchemy/sqlalchemy#10867] [Usage Questions]
I've learned that it is 'best practice' to use the SQLAlchemy built-in types for portability reasons to define your models.
I've been using the ` sqlalchemy.types.Text` type to define columns. This maps to the PostgreSQL `TEXT` datatype, the SQLite `TEXT` datatype and the MySQL `TEXT` datatype.

My problem is that, where on PostgreSQL the TEXT datatype is [defined](https://www.postgresql.org/docs/current/datatype-character.html#DATATYPE-CHARACTER) as 'variable unlimited length', on MySQL this is only maximum 2 ** 16 bytes which is 64k of data. And if I want to use a `LONGTEXT` in MySQL, which is the equivalent of the PostgreSQL `TEXT`, I have to use `from sqlalchemy.dialects.mysql import LONGTEXT`. Or, alternatively, I can define TEXT with a really large number such ` body = db.Column(db.Text(10000000)` as done in [this StackOverflow answer](https://stackoverflow.com/questions/64032326/convert-data-type-of-table-to-mediumtext-or-longtext-from-flask-sqlalchemy), but that feels like a hack.

This would mean that there is no 'neat' way to define a column in my model that is a `TEXT` in PostgreSQL and a `LONGTEXT` in MySQL.

I strongly feel that `sqlalchemy.types.Text` should actually map to `LONGTEXT` on MySQL. What am I missing here?

**Comment by zzzeek** [comment]
hi -

for special per-backend type condierations  use the [with_variant](https://docs.sqlalchemy.org/en/20/core/type_basics.html#types-with-variant) facility

```py
from sqlalchemy.dialects.mysql import LONGTEXT

Column('q', Text().with_variant(LONGTEXT, "mysql", "mariadb"))
```

there is no fully unlengthed type on the MySQL/MariaDB backends so there's not really any particular equivalent of PG's TEXT type.

**Reply by mbeijen** [reply]
While I understand that the MySQL LONGTEXT still is 'limited' to maximum 4GB (2^32) so it is not really 'unlimited', while the PG 'TEXT' is described as 'unlimited lenght' it seems like it is also limited to 1G of data [as described on its wiki](https://wiki.postgresql.org/wiki/FAQ#What_is_the_maximum_size_for_a_row.2C_a_table.2C_and_a_database.3F). This is also the default maximum for SQLites `TEXT`, as defined in [SQLITE_MAX_LENGTH](https://www.sqlite.org/limits.html)

And I'm not per se looking to write 'give me the PostgreSQL TEXT field on MySQL' , because I do understand that implementation details differ. I just want a way to define my column as 'give me the large text field on your database' which is TEXT on PostgreSQL and LONGTEXT on MySQL.

**Reply by zzzeek** [reply]
with_variant() will get you that in my answer above
